lntest/itest/lnd_forward_interceptor_test.go

```diff
@@ -0,0 +1,389 @@
+// +build rpctest
+
+package itest
+
+import (
+	"context"
+	"encoding/hex"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/btcsuite/btcd/wire"
+	"github.com/btcsuite/btcutil"
+	"github.com/lightningnetwork/lnd"
+	"github.com/lightningnetwork/lnd/lnrpc"
+	"github.com/lightningnetwork/lnd/lnrpc/routerrpc"
+	"github.com/lightningnetwork/lnd/lntest"
+	"github.com/lightningnetwork/lnd/routing/route"
+)
+
+type interceptorTestCase struct {
+	amountMsat        int64
+	invoice           *lnrpc.Invoice
+	shouldHold        bool
+	interceptorAction routerrpc.ResolveHoldForwardAction
+}
+
+// testForwardInterceptor tests the forward interceptor RPC layer.
+// The test creates a cluster of 3 connected nodes: Alice -> Bob -> Carol
+// Alice sends 4 different payments to Carol while the interceptor handles
+// differently the htlcs.
+// The test ensures that:
+// 1. Intercepted failed htlcs result in no payment (invoice is not settled).
+// 2. Intercepted resumed htlcs result in a payment (invoice is settled).
+// 3. Intercepted held htlcs result in no payment (invoice is not settled).
+// 4. When Interceptor disconnects it resumes all held htlcs, which result in
+//    valid payment (invoice is settled).
+func testForwardInterceptor(net *lntest.NetworkHarness, t *harnessTest) {
+
+	// initialize the test context with 3 connected nodes.
+	testContext := newInterceptorTestContext(t, net)
+	defer testContext.shutdownNodes()
+
+	const (
+		chanAmt = btcutil.Amount(300000)
+	)
+
+	// open and wait for channels.
+	testContext.openChannel(testContext.alice, testContext.bob, chanAmt)
+	testContext.openChannel(testContext.bob, testContext.carol, chanAmt)
+	defer testContext.closeChannels()
+	testContext.waitForChannels()
+
+	// connect the interceptor.
+	ctx := context.Background()
+	ctxt, cancelInterceptor := context.WithTimeout(ctx, defaultTimeout)
+	interceptor, err := testContext.bob.RouterClient.HtlcInterceptor(ctxt)
+	if err != nil {
+		t.Fatalf("failed to create HtlcInterceptor %v", err)
+	}
+
+	// prepare the test cases.
+	testCases, err := testContext.prepareTestCases()
+	if err != nil {
+		t.Fatalf("failed to prepare test cases")
+	}
+
+	// a channel for the interceptor go routine to send the requested packets.
+	interceptedChan := make(chan *routerrpc.ForwardHtlcInterceptRequest,
+		len(testCases))
+
+	// run the interceptor loop in its own go routine.
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for {
+			request, err := interceptor.Recv()
+			if err != nil {
+				return
```

Comment:
Any other action needed here? Log or fail test?

Reply:
In case this error is a result of context cancellation, nothing is needed as this is the test behavior. In other error cases it is better to fail the test. I added the code to support this behavior.

Reply:
To problem is that calling `t.Fatalf` isn't allowed in a goroutine

Reply:
Changed to `Errorf`